src/ui/public/vis/update_status.ts

```diff
@@ -18,61 +18,43 @@
  */
 
 import { calculateObjectHash } from './lib/calculate_object_hash';
-
-const Status = {
-  AGGS: 'aggs',
-  DATA: 'data',
-  PARAMS: 'params',
-  RESIZE: 'resize',
-  TIME: 'time',
-  UI_STATE: 'uiState',
-};
-
-// adapted from https://github.com/isaacs/json-stringify-safe/blob/02cfafd45f06d076ac4bf0dd28be6738a07a72f9/stringify.js
-function serializer() {
-  const stack = [];
-  const keys = [];
-
-  const cycleReplacer = function (key, value) {
-    if (stack[0] === value) return '[Circular ~]';
-    return `[Circular ~.${keys.slice(0, stack.indexOf(value)).join('.')}]`;
-  };
-
-  return function (key, value) {
-    if (stack.length > 0) {
-      const thisPos = stack.indexOf(this);
-      ~thisPos ? stack.splice(thisPos + 1) : stack.push(this);
-      ~thisPos ? keys.splice(thisPos, Infinity, key) : keys.push(key);
-      if (~stack.indexOf(value)) value = cycleReplacer.call(this, key, value);
-    }
-    else stack.push(value);
-
-    return value;
-  };
+import { Vis } from './vis';
+
+enum Status {
+  AGGS = 'aggs',
+  DATA = 'data',
+  PARAMS = 'params',
+  RESIZE = 'resize',
+  TIME = 'time',
+  UI_STATE = 'uiState',
 }
 
-function getUpdateStatus(requiresUpdateStatus, obj, param) {
+function getUpdateStatus<T extends Status>(
+  requiresUpdateStatus: T[] = [],
+  obj: any,
+  param: { vis: Vis; visData: any; uiState: any }
+): { [reqStats in T]: boolean } {
+  const status = {} as { [reqStats in T]: boolean };
 
   // If the vis type doesn't need update status, skip all calculations
-  if (!requiresUpdateStatus) {
-    return {};
+  if (requiresUpdateStatus.length === 0) {
+    return status;
   }
 
   if (!obj._oldStatus) {
     obj._oldStatus = {};
   }
 
-  const hasChangedUsingGenericHashComparison = (name, value) => {
-    const currentValue = JSON.stringify(value, serializer());
-    if (currentValue !== obj._oldStatus[name]) {
-      obj._oldStatus[name] = currentValue;
+  const hasChangedUsingGenericHashComparison = (name: string, value: any) => {
+    const hash = calculateObjectHash(value);
+    if (hash !== obj._oldStatus[name]) {
+      obj._oldStatus[name] = hash;
       return true;
     }
     return false;
   };
 
-  const hasSizeChanged = (currentWidth, currentHeight) => {
-
+  const hasSizeChanged = (currentWidth: number, currentHeight: number) => {
```

Comment:
Same for the previous comment. Prefer pure functions and move them outside the current function scope.
If not possible, at least change the function name to reflect that it not only check the status but also update the oldStatus.